lib/jekyll/document.rb

```diff
@@ -186,7 +186,8 @@ def url_placeholders
         path:        cleaned_relative_path,
         output_ext:  output_ext,
         name:        Utils.slugify(basename_without_ext),
-        title:       Utils.slugify(data['slug']) || Utils.slugify(basename_without_ext),
+        title:       Utils.slugify_cased(data['slug']) || Utils.slugify_cased(basename_without_ext),
```

Comment:
@parkr named parameter is a better solution, thanks for advice! 

BTW, you may want to take a look at issue #4135 , users are complaining about their underscore are replaced by hyphen, maybe someday users would say they want their `~` back. Should we use `Pretty` mode here for `:title`, seems Pretty mode `[^[:alnum:]._~!$&'()+,;=@]+` is designed to handle this kind of issue.